File: Samples/Verbum/how-to-highlight-all-found-word-entries-in-a-document-t202228-14.2.4-/VB/RichEdit/CustomSearchForm.vb
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq
Imports System.Windows.Forms
Imports DevExpress.XtraRichEdit
Imports DevExpress.XtraRichEdit.API.Native
Imports DevExpress.XtraRichEdit.Layout.Export

Namespace RichEdit
    Friend Class CustomSearchForm
        Inherits DevExpress.XtraRichEdit.Forms.SearchTextForm

        Public Sub New(ByVal parameters As DevExpress.XtraRichEdit.Forms.SearchFormControllerParameters)
            MyBase.New(parameters)
            AddHandler Me.cbFndSearchString.EditValueChanged, AddressOf cbFndSearchString_EditValueChanged
            AddHandler Me.chbFndMatchCase.CheckedChanged, AddressOf chbFndMatchCase_CheckedChanged
            AddHandler Me.chbFndFindWholeWord.CheckedChanged, AddressOf chbFndMatchCase_CheckedChanged
            AddHandler Me.FormClosed, AddressOf CustomSearchForm_FormClosed
        End Sub

        Private Sub chbFndMatchCase_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
            Me.cbFndSearchString.EditValue = Nothing
        End Sub

        Private Sub CustomSearchForm_FormClosed(ByVal sender As Object, ByVal e As FormClosedEventArgs)
            ClearCustomMarks()
            RemoveHandler Me.cbFndSearchString.EditValueChanged, AddressOf cbFndSearchString_EditValueChanged
            RemoveHandler Me.chbFndMatchCase.CheckedChanged, AddressOf chbFndMatchCase_CheckedChanged
            RemoveHandler Me.chbFndFindWholeWord.CheckedChanged, AddressOf chbFndMatchCase_CheckedChanged
        End Sub
        Private Sub cbFndSearchString_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            ClearCustomMarks()
            If Me.cbFndSearchString.EditValue IsNot Nothing Then
                Dim text As String = Me.cbFndSearchString.EditValue.ToString()
                If Not String.IsNullOrEmpty(text) Then
                    CreateCustomMarks(text)
                End If
            End If
        End Sub
        Private Sub CreateCustomMarks(ByVal text As String)
            Dim options As DevExpress.XtraRichEdit.API.Native.SearchOptions = DevExpress.XtraRichEdit.API.Native.SearchOptions.None
            If Controller.FindWholeWord Then
                options = options Or DevExpress.XtraRichEdit.API.Native.SearchOptions.WholeWord
            End If
            If Controller.CaseSensitive Then
                options = options Or DevExpress.XtraRichEdit.API.Native.SearchOptions.CaseSensitive
            End If
            Dim ranges() As DocumentRange = Control.Document.FindAll(text, options)
            If ranges IsNot Nothing AndAlso ranges.Length > 0 Then
                For Each range As DocumentRange In ranges
                    Control.Document.CustomMarks.Create(range.Start, range)
                    Control.Document.CustomMarks.Create(range.End, range)
                Next range
            End If
        End Sub
        Private Sub ClearCustomMarks()
            For i As Integer = Control.Document.CustomMarks.Count - 1 To 0 Step -1
                Dim mark As DevExpress.XtraRichEdit.API.Native.CustomMark = Control.Document.CustomMarks(i)
                Control.Document.CustomMarks.Remove(mark)
            Next i
        End Sub
    End Class
End Namespace

File: Samples/Verbum/how-to-highlight-all-found-word-entries-in-a-document-t202228-14.2.4-/CS/RichEdit/CustomSearchForm.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;
using DevExpress.XtraRichEdit;
using DevExpress.XtraRichEdit.API.Native;
using DevExpress.XtraRichEdit.Layout.Export;

namespace RichEdit {
    class CustomSearchForm : DevExpress.XtraRichEdit.Forms.SearchTextForm {
        public CustomSearchForm(DevExpress.XtraRichEdit.Forms.SearchFormControllerParameters parameters)
            : base(parameters) {
            this.cbFndSearchString.EditValueChanged += cbFndSearchString_EditValueChanged;
            this.chbFndMatchCase.CheckedChanged += chbFndMatchCase_CheckedChanged;
            this.chbFndFindWholeWord.CheckedChanged += chbFndMatchCase_CheckedChanged;
            this.FormClosed += CustomSearchForm_FormClosed;
        }

        void chbFndMatchCase_CheckedChanged(object sender, EventArgs e) {
            this.cbFndSearchString.EditValue = null;
        }

        void CustomSearchForm_FormClosed(object sender, FormClosedEventArgs e) {
            ClearCustomMarks();
            this.cbFndSearchString.EditValueChanged -= cbFndSearchString_EditValueChanged;
            this.chbFndMatchCase.CheckedChanged -= chbFndMatchCase_CheckedChanged;
            this.chbFndFindWholeWord.CheckedChanged -= chbFndMatchCase_CheckedChanged;
        }
        void cbFndSearchString_EditValueChanged(object sender, EventArgs e) {
            ClearCustomMarks();
            if(this.cbFndSearchString.EditValue != null) {
                string text = this.cbFndSearchString.EditValue.ToString();
                if(!String.IsNullOrEmpty(text))
                    CreateCustomMarks(text);
            }
        }
        void CreateCustomMarks(string text) {
            DevExpress.XtraRichEdit.API.Native.SearchOptions options = DevExpress.XtraRichEdit.API.Native.SearchOptions.None;
            if(Controller.FindWholeWord)
                options |= DevExpress.XtraRichEdit.API.Native.SearchOptions.WholeWord;
            if(Controller.CaseSensitive)
                options |= DevExpress.XtraRichEdit.API.Native.SearchOptions.CaseSensitive;
            DocumentRange[] ranges = Control.Document.FindAll(text, options);
            if(ranges != null && ranges.Length > 0) {
                foreach(DocumentRange range in ranges) {
                    Control.Document.CustomMarks.Create(range.Start, range);
                    Control.Document.CustomMarks.Create(range.End, range);
                }
            }
        }
        void ClearCustomMarks() {
            for(int i = Control.Document.CustomMarks.Count - 1; i >= 0; i--) {
                DevExpress.XtraRichEdit.API.Native.CustomMark mark = Control.Document.CustomMarks[i];
                Control.Document.CustomMarks.Remove(mark);
            }
        }
    }
}
